### PR TITLE
Add expanded mount format to stack deploy

### LINKED ISSUE #29193: Support expanded ports and mount format in the the compose v3 format
Following the decisions made in #28527 we should also support the expanded ports and mount format. Instead of CSV we should use an object for each.

### LINKED ISSUE #31224: Compose file parsing can't parse Windows paths
**Description**

The compose file parsing inside Docker can't parse Windows volumes.

Just opening this to track and in case anyone else was experimenting with Swarm and Windows.

Tracking the status of: https://github.com/docker/docker/blob/master/cli/compose/convert/volume.go#L31

**Steps to reproduce the issue:**
1. Create a v3 docker-compose.yml spec with a service.volume spec.
2. Run `docker stack deploy...` 

**Describe the results you received:**
`undefined volume: C`

**Describe the results you expected:**

Volume mounts in a container

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 1.13.1
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: active
 NodeID: ov0y3pqc1gjx15upv9mjru6bv
 Is Manager: true
 ClusterID: vkznxg9rw3dmput8jl9k291fz
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.16.58.181
 Manager Addresses:
  172.16.58.181:2377
Default Isolation: hyperv
Kernel Version: 10.0 15031 (15031.0.amd64fre.rs2_release.170204-1546)
Operating System: Windows 10 Enterprise Insider Preview
OSType: windows
Architecture: x86_64
CPUs: 1
Total Memory: 3 GiB
Name: DESKTOP-UJ1M2UA
ID: 5JUS:Y5TR:DAEX:KQ7U:SC3Z:3IBV:3HBH:XNN5:K7SP:JDL6:J2OB:6FQR
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 131
 System Time: 2017-02-21T08:17:39.3966974-08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

### PR DESCRIPTION
**Pull request by dnephin**
Fixes  #29193 
Fixes #31224
Also (should) fix some windows support for volumes

I've added this to Compose format v3.1 for now, but it might be more appropriate for a v3.2, depending on when it gets merged and where it gets cherry-picked.

@cpuguy83 I'd appreciate your review of 3c74323d94491571e8667ba0a4b2cca9911b4dc0. It adds a client-side parser for the old volume spec. As far as I know this is the first time we're attempting to fully parse the spec on the client (without being aware of the engine platform). It assumes no knowledge of the platform, so windows and linux container paths are both supported on all clients.

@shin- we will probably need to implement this on the python side as well. It should be a smaller change. I think we'll only need to copy the schema and implement something like 3c74323d94491571e8667ba0a4b2cca9911b4dc0

cc @vdemeester

### CODE REVIEW

**Review by vdemeester — COMMENTED** [review]
SGTM 👼
I wonder how using `volume` fits with ongoing discussion on #28527 (volumes, mounts ?).

/cc @icecrime @stevvooe

File: cli/compose/loader/loader_test.go [hunk 1]
```diff
@@ -825,3 +825,31 @@ networks:
...
 	assert.Equal(t, expected, config.Networks)
 }
+
+func TestLoadExpandedMountFormat(t *testing.T) {
+	config, err := loadYAML(`
+version: "3.1"
+services:
+  web:
+    image: busybox
+    volumes:
```
Comment (stevvooe) [inline cli/compose/loader/loader_test.go]: Shouldn't these be called mounts? It is very confusing to have a volume declaration here then a volume declaration at the top level. This contributes to confusion. Not that this section will also handle binds, so calling it volumes is inaccurate.

**Comment by stevvooe** [comment]
The format looks good, in general. I would really prefer that we call it mounts, or at least have that as an alias. It is very weird to have `volumes`, then `type: volume`. It is also problematic to mix binds, volume, tmpfs, (and soon image mounts!) into volume. We should focus the term volume around the orchestratable object.

File: cli/compose/loader/loader_test.go (hunk 1 above)
Comment (dnephin) [inline cli/compose/loader/loader_test.go]: Yes, but it's also a breaking change, so not something we can change in the `v3.1` format. We missed our chance with v3, so we can queue it up for v4 I guess (depending on the outcome of the `docker run` flag change).

**Comment by dnephin** [comment]
We could do a `mount` alias

**Comment by dnephin** [comment]
Rebased

File: cli/compose/schema/schema.go [hunk 2]
```diff
@@ -113,23 +117,45 @@ func humanReadableType(definition string) string {
...
 		}
 	}
 
-	return mostSpecificError
+	if mostSpecificError+1 == len(errors) {
+		return validationError{parent: errors[mostSpecificError]}
+	}
+
+	switch errors[mostSpecificError].Type() {
+	case "number_one_of", "number_any_of":
```
Comment (AkihiroSuda) [inline cli/compose/schema/schema.go]: Can we make this strings constant values?

Comment (dnephin) [inline cli/compose/schema/schema.go]: sure

**Comment by justincormack** [comment]
cc @cpuguy83 @jhowardmsft

**Comment by cpuguy83** [comment]
Spoke with Daniel since I was having some issues trying to get this going. From a functionality POV, just have a couple of issues:

- When "driver" is not defined on the volume, "driver_opts" are not passed through
- The `--mount` cli supports templating everything about the mount spec, including `Source` which allows for some really powerful use-cases... this is kind of lost here afaict.

**Comment by dnephin** [comment]
> The --mount cli supports templating everything about the mount spec, including Source which allows for some really powerful use-cases... this is kind of lost here afaict.

I believe it's still possible for all fields. For `source` it requires that you use an external volume, because the source name in the service has to be a reference to a volume definition. The volume definition can use a template with:

```yaml
data:
  external:
    name: '{{ .Template }}'

```

**Comment by dnephin** [comment]
ping

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by vdemeester — COMMENTED** [review]
Should we add a `mount` alias though ? 👼

**Comment by dnephin** [comment]
When it's supported by the CLI sure, but I think that's not required to merge this. We can add it later.

**Review by vdemeester — APPROVED** [review]
Sounds fair.
LGTM 🦁

**Comment by thaJeztah** [comment]
@dnephin can you open a pull request in the https://github.com/docker/docker.github.io repository for the syntax changes in the docker-compose file?

**Comment by dnephin** [comment]
Opened https://github.com/docker/docker.github.io/pull/2283

**Comment by cpuguy83** [comment]
@dnephin what cli? Service is --mount as is.

**Comment by dnephin** [comment]
True. The compose file fields are named after the `docker run` cli. Either way we can't drop `volumes` as the key in a minor version, so we can either decide to add an alias of `mounts` during code freeze, or we can decide to worry about it for the next version.

**Comment by justincormack** [comment]
@dnephin is there any reason we can't do a major bump of compose format for 17.04?

**Comment by dnephin** [comment]
You mean make it v4? There are no backwards incompatible changes, or any significantly large changes, so v4 would not be appropriate.

**Comment by justincormack** [comment]
@dnephin I meant so we can add `mounts` which you said was not suitable for a minor bump.

**Comment by justincormack** [comment]
Ok, lets leave any `mounts` change for another PR that bumps the API.

**Comment by dnephin** [comment]
I see what you mean. I think we'll probably want to batch up a few other breaking changes all at once before making a v4.

**Pull request merged**